### PR TITLE
OOPIF full support in extension

### PR DESCRIPTION
**Pull request by patrickhulce**
broken out from #7517 for easier diffing

**tl;dr** I don't see an immediate way to make extensions work with `setAutoAttach`

**Summary**
The problem: [extension debugger API](https://developers.chrome.com/extensions/debugger#method-sendCommand) does not have a parameter for `sessionId`. It's method of talking to targets uses `targetId` and a different `sendCommand` just to them. This PR implements that approach, *however*, we run into a problem.

1. You cannot `chrome.debugger.sendCommand` to a target you have not yet attached to.
2. You cannot `chrome.debugger.attach` to a target that has already been attached to by another debugger (which `Target.setAutoAttach` seems to do).

The options as I see it:

1. Get `chrome.debugger.sendCommand` to support `sessionId` as a `Debuggee` target (maybe there's hidden support I haven't found?? I tried sending it as part of params too with no luck)
2. Switch back to `discover` all the targets and filter the frame tree instead of `setAutoAttach`
3. Something I'm missing?

**Related Issues/PRs**
#7517

### CODE REVIEW

**Comment by paulirish** [comment]
talked with pavel and dmitry and they think setAutoAttach can work. (without it, we have a race)

### Introducing @pavelfeldman's Foolproof Guide to Making `setAutoAttach` Work with the Extension API

1. Target.setAutoAttach({setAutoAttach:true, waitForDebuggerOnStart: true})
1. targetCreated event arrives. 
1. attachedToTarget event arrives.
   - attachedToTarget has sessionID, targetinfo.targetID
1. newdebuggee = chrome.debugger.attachtoTarget(targetID)
1. newdebugeee: set up all agents, overrides
1. main connection: Target.detachFromTarget(sessionID)


@patrickhulce since you've been deep in this, does it sound reasonable to you?

**Comment by brendankenny** [comment]
is it possible there will be a significant impact from `waitForDebuggerOnStart: true`, or are we connecting quickly enough that it's ok?

**Comment by paulirish** [comment]
>is it possible there will be a significant impact from waitForDebuggerOnStart: true, or are we connecting quickly enough that it's ok? 

without using it we can very likely miss out on network activity that happened while we weren't attached. (probably for a datauri iframe).
with pausing yes we wait for our JS thread to respond and attach, but I'd expect this to happen in <5ms. So it does add some latency, but probably not a lot. Our thread _shouldnt_ be busy at this time.

**Comment by patrickhulce** [comment]
@paulirish does Pavel think the `waitForDebuggerOnStart` will fix the fact that the chrome extension throws when you try to attach to a target that was discovered via `setAutoAttach`? 

That was the core problem I was trying to communicate that we're in a catch 22. The extension API wants you to connect to a target before sending a command yet when you try to attach it says the target is already attached (I'm assuming because it was attached via `setAutoAttach` but possibly it's another reason).

**Comment by patrickhulce** [comment]
Aside from `waitForDebuggerOnStart` I've implemented exactly as he has described, I'll give it a shot.

**Comment by patrickhulce** [comment]
The plot thickens, you can't use `Target.detachFromTarget` at all from extension 🤔 

> Protocol error (Target.detachFromTarget): Not allowed

This is the core of the problem AFAICT: extension API just can't interact with sessionId-based attached targets.

- It can't sendCommand based on a sessionId (this is based on the docs and everywhere I've tried to stick `sessionId` to no effect or error)
- It throws when attaching to a targetId that has been auto-attached (via `chrome.debugger.attach`)
- It throws when attaching to a targetId via `Target.attachToTarget` (pretty sure I tried this earlier and it threw you have to go through `chrome.debugger.attach`)
- It throws when detaching from a sessionId that has been auto-attached (via `Target.detachFromTarget`, see above)
- It can't detach via `targetId` (via `chrome.debugger.detach`) because it doesn't think it's attached in the first place.

**Comment by paulirish** [comment]
+ @dgozman @pavelfeldman

**Comment by pavelfeldman** [comment]
It looks like our `Target` domain handler for `chrome.debugger` endpoints is [constrained](https://cs.chromium.org/chromium/src/content/browser/devtools/render_frame_devtools_agent_host.cc?sq=package:chromium&g=0&l=300) due to security considerations. It will auto-attach, but it won't let you attach or detach :/

Which means that magic that I was suggesting does not work, sorry.

**Comment by patrickhulce** [comment]
So it seems like our options are....

- Use `setAutoAttach` and don't support OOPIFs in the extension
- Use `discoverTargets` and live with potential races
- Implement both methods and choose which one we use based on some environment flag
- Other??

**Comment by pavelfeldman** [comment]
> Use discoverTargets and live with potential races

Can't do even that. ([link](https://cs.chromium.org/chromium/src/content/browser/devtools/protocol/target_handler.cc?sq=package:chromium&g=0&l=554))

**Comment by pavelfeldman** [comment]
Remind me why we don't want to use autoAttach / sendMessageToTarget?

**Comment by patrickhulce** [comment]
> Use discoverTargets and live with potential races
> Can't do even that. (link)

@pavelfeldman yes, it would never be both. I know y'all have been recommending against `discoverTargets` from the beginning but it is technically *an* option :)

> Remind me why we don't want to use autoAttach / sendMessageToTarget?

Extension API throws when you try to use `sendMessageToTarget` and tells you to use `chrome.debugger.sendCommand` which has the aforementioned attachment problems

**Comment by pavelfeldman** [comment]
>  it is technically an option

not in chrome.debugger land, right?

**Comment by pavelfeldman** [comment]
> Extension API throws when you try to use `sendMessageToTarget`

I'll look into this one.

**Comment by patrickhulce** [comment]
> not in chrome.debugger land, right?

I haven't implemented it but it seems like it should work from there, no or am I missing something? I can still see and attach to targets when they aren't auto attached so I should be able to attach to *them*, right?

**Comment by patrickhulce** [comment]
> I'll look into this one.

To clarify, our current method when it's not `flat` works for `sendMessageToTarget` it only started failing when I started using the new method

**Comment by pavelfeldman** [comment]
> haven't implemented it but it seems like it should work from there

I sent you a chrome link above that suggests there is no way it does

**Comment by patrickhulce** [comment]
> Error: Protocol error (Target.sendMessageToTarget): When using flat protocol, messages are routed to the target via the sessionId attribute.

That's what I get once I enable `flatten`

**Comment by pavelfeldman** [comment]
> To clarify, our current method when it's not flat works for `sendMessageToTarget`

`sendMessageToTarget` is not used in flat mode at all. So it is fine for it to throw when misused with flat. But if non-flat works, why touching it?

**Comment by patrickhulce** [comment]
> I sent you a chrome link above that suggests there is no way it does

Oh I misunderstood I thought only the subtargets were kAutoAttachOnly but now I see.

**Comment by patrickhulce** [comment]
> But if non-flat works, why touching it?

Because we want to do this for all descendants not just children. I guess you're saying we can continue to use the old way if we just properly nest the `Target.sendMessageToTarget` commands enough?

**Comment by pavelfeldman** [comment]
> I guess you're saying we can continue to use the old way if we just properly nest the `Target.sendMessageToTarget` commands enough

Exactly this

**Comment by patrickhulce** [comment]
Alrighty we've found our solution even if it means we can't use any of the flattening niceness

Thanks @pavelfeldman !
